docs/providers/provider-protocol-reference.md

```diff
@@ -0,0 +1,243 @@
+
+register_custom_getrandom!(use_fp_bindings);
+```
+
+> Note: this example snippet might prove generic and useful enough to be upstreamed
+  by Fiberplane and available as a public helper crate. Stay tuned!
+
+### `now`
+
+#### Signature
+
+```rust
+pub fn now() -> Timestamp;
+```
+
+#### Description
+
+Returns a timestamp of the current time of execution. The usefulness of this
+timestamp comes from what you want to do with timers, or signature
+creation/verification schemes.
+
+## Exported functions
+
+Exported functions are functions that your provider implements, and that the
+runtime will call. In order to make those exported functions usable in Fiberplane,
+they must be annotated with `#[pdk_export]`.
+
+### `get_supported_query_types`
+
+It is usually safer to use the Plugin Development Kit macros to let the compiler
+generate correct implementations of these functions. A correct `pdk_query_types!`
+invocation will generate this function for you. See the
+[tutorial](doc:create-a-provider) for more information about using the PDK
+
+#### Signature
+
+```rust
+pub async fn get_supported_query_types(config: ProviderConfig) -> Vec<SupportedQueryType>;
+```
+
+#### Description
+
+Returns the query types supported by this provider.
+This function allows Studio to know upfront which formats will be
+supported, and which providers (and their query types) are eligible to
+be selected for certain use cases.
+
+### `invoke2`
+
+It is usually safer to use the Plugin Development Kit macros to let the compiler
+generate correct implementations of these functions. A correct `pdk_query_types!`
+invocation will generate this function for you. See the
+[tutorial](doc:create-a-provider) for more information about using the PDK
+
+#### Signature
+
+```rust
+pub async fn invoke2(request: ProviderRequest) -> Result<Blob>;
+```
+
+#### Description
+
+Invokes the provider to perform a data request.
+
+### `create_cells` (Optional)
+
+#### Signature
+
+```rust
+pub fn create_cells(query_type: String, response: Blob) -> Result<Vec<Cell>>;
+```
+
+#### Description
+
+Creates output cells based on the response (a `Blob` that the provider
+previously sent to answer an `invoke2` call).
+Studio would typically embed the created cells in the provider cell,
+but other actions could be desired.
+
+When any created cells use a `data-links` field with the value
+`cell-data:<mime-type>,self` (such as Log cells or Graph cells),
+Studio will replace the value `self` with
+the ID of the cell for which the query was invoked. This allows the
+provider to create cells that reference its own data without knowing the
+context of the cell in which it was executed.
+
+Note: When the MIME type in the provider response is
+`application/vnd.fiberplane.cells` (suffixed with either `+json` or
+`+msgpack`), Studio will elide the call to `create_cells()` and simply
+parse the data directly to a `Vec<Cell>`. This is what makes
+`create_cells`'s implementation optional.
```

Comment:
```suggestion
Note that implementing `create_cells` is optional. If `invoke2` returns a Blob with a MIME type
of `application/vnd.fiberplane.cells` (suffixed with either `+json` or `+msgpack`), that response
will be parsed into an array of cells, and the call to this function is elided. A consequence of
this approach is that the blob with the response does not get stored on the provider cell --
only the parsed cells will be stored. This can make this approach less suitable for graph cells
and log cells, for instance, but makes it more efficient when there is no need to refer to the
data after the cells are instantiated.
```